js/modules/k6/experimental/fs/module.go

```diff
@@ -0,0 +1,156 @@
+// Package fs provides a k6 module that allows users to interact with files from the
+// local filesystem.
+package fs
+
+import (
+	"fmt"
+
+	"github.com/dop251/goja"
+	"go.k6.io/k6/js/modules"
+	"go.k6.io/k6/lib/fsext"
+)
+
+// TODO @oleiade: we need to explicitly document that File are read-only and are really views, thus
+// any changes made to the file from outside of k6, while it runs, will be ignored as k6 will work
+// on its own cached version of the file's content.
+
+type (
+	// RootModule is the global module instance that will create instances of our
+	// module for each VU.
+	RootModule struct {
+		fileRegistry fileRegistry
+	}
+
+	// ModuleInstance represents an instance of the fs module for a single VU.
+	ModuleInstance struct {
+		vu modules.VU
+
+		fileRegistry *fileRegistry
+	}
+)
+
+var (
+	_ modules.Module   = &RootModule{}
+	_ modules.Instance = &ModuleInstance{}
+)
+
+// New returns a pointer to a new [RootModule] instance.
+func New() *RootModule {
+	return &RootModule{
+		fileRegistry: fileRegistry{
+			files: make(map[string]*[]byte),
+		},
+	}
+}
+
+// NewModuleInstance implements the modules.Module interface and returns a new
+// instance of our module for the given VU.
+func (rm *RootModule) NewModuleInstance(vu modules.VU) modules.Instance {
+	return &ModuleInstance{vu: vu, fileRegistry: &rm.fileRegistry}
+}
+
+// Exports implements the modules.Module interface and returns the exports of
+// our module.
+func (mi *ModuleInstance) Exports() modules.Exports {
+	return modules.Exports{
+		Named: map[string]any{
+			"open":     mi.Open,
+			"File":     file{},
+			"FileInfo": fileInfo{},
+		},
+	}
+}
+
+// Open opens a file and returns a promise that will resolve to a [File] instance
+func (mi *ModuleInstance) Open(path goja.Value) *goja.Promise {
+	rt := mi.vu.Runtime()
+
+	promise, resolve, reject := makeHandledPromise(mi.vu)
```

Comment:
As part of that goja abstraction layer we discussed, what I meant was that we could have these user-facing methods wrap internal methods that actually have the implementation. This way we can avoid interacting with goja in most of our tests, which would simplify things, and then have just a few goja-level tests to ensure that the abstraction layer is working properly.

So in this case we could have an internal `open()` method that takes `pathStr` and works with `FileSystems["file"]`, without any of the promises and async behavior. And in this user-facing `Open()` we can do the `goja.Value` extraction and async implementation that calls the internal `open()`. This can then possibly be reused for other user-facing methods as well.

Reply:
I'll give it a shot 👍🏻